cmd/fetchd/cmd/genasiupgrade.go

```diff
@@ -73,16 +98,82 @@ func ASIGenesisUpgradeCmd(defaultNodeHome string) *cobra.Command {
 	return cmd
 }
 
-func ASIGenesisUpgradeReplaceDenomMetadata() {}
+func ASIGenesisUpgradeReplaceDenomMetadata(cdc codec.Codec, appState *map[string]json.RawMessage) error {
+	bankGenState := banktypes.GetGenesisStateFromAppState(cdc, *appState)
+
+	OldBaseDenomUpper := strings.ToUpper(OldBaseDenom)
+	NewBaseDenomUpper := strings.ToUpper(NewBaseDenom)
+
+	denomRegex := getRegex(OldDenom, NewDenom)
+	upperDenomRegex := getRegex(OldBaseDenomUpper, NewBaseDenomUpper)
+	exponentDenomRegex := getPartialRegexLeft(OldBaseDenom, NewBaseDenom)
+
+	for _, metadata := range bankGenState.DenomMetadata {
+		replaceString(&metadata.Base, []*regexPair{denomRegex})
+		if metadata.Name == OldBaseDenomUpper {
+			metadata.Description = NewDescription
+			metadata.Display = NewBaseDenomUpper
+			metadata.Name = NewBaseDenomUpper
+			metadata.Symbol = NewBaseDenomUpper
+		}
+		for _, unit := range metadata.DenomUnits {
+			replaceString(&unit.Denom, []*regexPair{upperDenomRegex})
+			replaceString(&unit.Denom, []*regexPair{exponentDenomRegex})
+		}
+	}
+
+	bankGenStateBytes, err := cdc.MarshalJSON(bankGenState)
+	if err != nil {
+		return fmt.Errorf("failed to marshal auth genesis state: %w", err)
+	}
+
+	(*appState)[banktypes.ModuleName] = bankGenStateBytes
+	return nil
+}
 
 func ASIGenesisUpgradeReplaceChainID(genesisData *types.GenesisDoc) {
 	genesisData.ChainID = NewChainId
 }
 
-func ASIGenesisUpgradeReplaceDenom() {}
+func ASIGenesisUpgradeReplaceDenom(jsonString *string) {
+	for _, target := range []string{"denom", "bond_denom", "mint_denom", "base_denom", "base"} {
+		re := regexp.MustCompile(fmt.Sprintf(`("%s"\s*:\s*)"%s"`, target, OldDenom))
+		if re.MatchString(*jsonString) {
+			*jsonString = re.ReplaceAllString(*jsonString, fmt.Sprintf(`${1}"%s"`, NewDenom))
+		}
+	}
```

Comment:
Perhaps it would be better to use the same recursive iteration through the json map container. in that case we would not even need to use the regexp,, it si possibe to specify exactly the key and value - right now the key & value is part of regexp, while in JSON map contaner we can exactly access & compare both individually.